### PR TITLE
HDDS-8058. Gracefully handle invalid S3 authorization header.

### PR DESCRIPTION
**Pull request by SaketaChalamchala**
## What changes were proposed in this pull request?

Added a check in AWSV4AuthValidator.java to ensure that the String to Sign conforms to the AWS V4 format. 
If the authorization header is invalid or conforms to an older version an error would be thrown.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8058

## How was this patch tested?

Added a test in TestAWSV4AuthValidator.java
Also, Manually tested submitting S3 requests with AWS V4 and V2 headers in ozonesecure dev docker. See attached files for signature calculations.
V2 header : 
```
sh-4.2$ strToSign="PUT\n\n${content_type}\n${date}\n${resource}"
sh-4.2$ echo -en ${strToSign}
PUT

application/octet-stream
Fri, 10 Mar 2023 03:46:36 +0000
/bucket1/1.txt

sh-4.2$ signature=`echo -en ${strToSign} | openssl sha1 -hmac ${s3_secret} -binary | base64`
sh-4.2$ echo ${signature}
qadnu2/QbXgLIH7Z8G0Rs+rzDvA=

sh-4.2$ curl -X PUT -T "${file}" -H "Host: ${host}" -H "Date: ${date}" -H "Content-Type: ${content_type}" -H "Authorization: AWS ${s3_key}:${signature}" -w "%{http_code}" http://${host}:9878${resource}
<?xml version="1.0" encoding="UTF-8"?>
<Error>
  <Code>AccessDenied</Code>
  <Message>User doesn't have the right to access this resource.</Message>
  <Resource>1.txt</Resource>
  <RequestId>1c5e00c6-5aa1-4376-93a6-b82bc6f438ba</RequestId>
</Error>
403
```

V4 header : 
```
...
sh-4.2$ HTTP_CANONICAL_REQUEST="PUT
${HTTP_CANONICAL_REQUEST_URI}\n
${HTTP_CANONICAL_REQUEST_HEADERS}\n
${HTTP_REQUEST_SIGNED_HEADERS}
${HTTP_REQUEST_PAYLOAD_HASH}"

sh-4.2$ echo -en $HTTP_CANONICAL_REQUEST
PUT
/bucket1/1.txt

content-type:application/octet-stream
host::9878
x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date:20230310T205249Z

content-type;host;x-amz-content-sha256;x-amz-date
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855

sh-4.2$ stringToSign="AWS4-HMAC-SHA256\n${CURRENT_DATE_ISO8601}\n${CURRENT_DATE_DAY}/${AWS_REGION}/${AWS_SERVICE}/aws4_request\n$(hash_sha256 "${HTTP_CANONICAL_REQUEST}")"
sh-4.2$ echo $stringToSign
AWS4-HMAC-SHA256
20230310T205249Z
20230310/us-east-1/s3/aws4_request
34eab1496779bd03ed30a25f1a186c9b0af029f998f8a3ec12eb3872d3e831cd

...
sh-4.2$ # Create the signature.
sh-4.2$ function create_signature {
  stringToSign="AWS4-HMAC-SHA256\n${CURRENT_DATE_ISO8601}\n${CURRENT_DATE_DAY}/${AWS_REGION}/${AWS_SERVICE}/aws4_request\n$(hash_sha256 "${HTTP_CANONICAL_REQUEST}")"
  dateKey=$(hmac_sha256 key:"AWS4${AWS_SECRET_ACCESS_KEY}" "${CURRENT_DATE_DAY}")
  regionKey=$(hmac_sha256 hexkey:"${dateKey}" "${AWS_REGION}")
  serviceKey=$(hmac_sha256 hexkey:"${regionKey}" "${AWS_SERVICE}")
  signingKey=$(hmac_sha256 hexkey:"${serviceKey}" "aws4_request")
  
  printf "${stringToSign}" | openssl dgst -sha256 -mac HMAC -macopt hexkey:"${signingKey}" | sed 's/(stdin)= //'
}

sh-4.2$ SIGNATURE="$(create_signature)"
sh-4.2$ echo $SIGNATURE 
5e9f39c0fea0b6d5a9dad4f7ee900d0367491993f6047ffe485bc9750bb2ff25

sh-4.2$ HTTP_REQUEST_AUTHORIZATION_HEADER="\
AWS4-HMAC-SHA256 Credential=${AWS_ACCESS_KEY_ID}/${CURRENT_DATE_DAY}/\
${AWS_REGION}/${AWS_SERVICE}/aws4_request, \
SignedHeaders=${HTTP_REQUEST_SIGNED_HEADERS}, Signature=${SIGNATURE}"

sh-4.2$ IN_FILE=./1.txt

sh-4.2$ curl "http://${host}:9878${HTTP_CANONICAL_REQUEST_URI}" \
-H "Authorization: ${HTTP_REQUEST_AUTHORIZATION_HEADER}" \
-H "content-type: ${HTTP_REQUEST_CONTENT_TYPE}" \
-H "x-amz-content-sha256: ${HTTP_REQUEST_PAYLOAD_HASH}" \
-H "x-amz-date: ${CURRENT_DATE_ISO8601}" \
-T ${IN_FILE}

sh-4.2$ ozone fs -ls ofs://om/s3v/bucket1/
Found 2 items
drwxrwxrwx   - om om          0 2023-03-10 21:09 ofs://om/s3v/bucket1/.Trash
-rw-rw-rw-   3 om om         14 2023-03-10 21:11 ofs://om/s3v/bucket1/1.txt
```
Om_error_for_aws_v2
[aws_v2_sign_calc.txt](https://github.com/apache/ozone/files/10946447/aws_v2_sign_calc.txt)
[aws_v4_sign_calc.txt](https://github.com/apache/ozone/files/10946448/aws_v4_sign_calc.txt)

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java [hunk 1]
```diff
@@ -39,6 +40,11 @@ final class AWSV4AuthValidator {
   private static final Logger LOG =
       LoggerFactory.getLogger(AWSV4AuthValidator.class);
   private static final String HMAC_SHA256_ALGORITHM = "HmacSHA256";
+  private static final Pattern STR_TO_SIGN_PATTERN =
```
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java:43]: Can you add a sample doc for the string that need to be matched.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/security/TestAWSV4AuthValidator.java [hunk 2]
```diff
@@ -64,15 +67,25 @@ public static Collection<Object[]> data() {
                 "577efef23edd43b7e1a59",
             "5d672d79c15b13162d9279b0855cfba" +
                 "6789a8edb4c82c400e06b5924a6f2b5d7",
-            "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY"
+            "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY",
+            true
+        },
+        {
+            "PUT\n" +
```
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/security/TestAWSV4AuthValidator.java:74]: Can we add a few more test cases? Will the regex flag `AWS4-ZAVC-HJUA123` in the start of the header? The regular expression added is the main entity to check here.

**Comment by kerneltime** [comment]
cc @tanvipenumudy can you take a look ?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java [hunk 3]
```diff
@@ -39,6 +40,11 @@ final class AWSV4AuthValidator {
...
+          Pattern.compile("^AWS4[A-Z\\-\\d]+\n" +
+                  "\\d{8}T\\d{6}Z\n" +
+                  "\\d{8}/[a-z\\-\\d]+/[a-zA-Z\\d]+/aws4_request\n" +
+                  "[a-fA-F\\d]+$");
```
Comment (tanvipenumudy) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java:47]:
```suggestion
                  "[a-fA-F\\d]{64}$");
```

We could modify `[a-fA-F\\d]+` to `[a-fA-F\\d]{64}` to match exactly 64 characters of a hexadecimal string.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java [hunk 4]
```diff
@@ -39,6 +40,11 @@ final class AWSV4AuthValidator {
   private static final Logger LOG =
       LoggerFactory.getLogger(AWSV4AuthValidator.class);
   private static final String HMAC_SHA256_ALGORITHM = "HmacSHA256";
+  private static final Pattern STR_TO_SIGN_PATTERN =
+          Pattern.compile("^AWS4[A-Z\\-\\d]+\n" +
```
Comment (tanvipenumudy) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java:44]:
```suggestion
          Pattern.compile("^AWS4-HMAC-SHA256\n" +
```
We could add `AWS4-HMAC-SHA256` towards the beginning of the string to adhere by the AWS V4 signature format.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java [hunk 5]
```diff
@@ -39,6 +40,11 @@ final class AWSV4AuthValidator {
...
+                  "\\d{8}T\\d{6}Z\n" +
+                  "\\d{8}/[a-z\\-\\d]+/[a-zA-Z\\d]+/aws4_request\n" +
```
Comment (tanvipenumudy) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/AWSV4AuthValidator.java:46]:
```suggestion
                  "\\d{8}/[a-z\\-\\d]+/[a-z\\d]+/aws4_request\n" +
```
I think we could only consider to include lowercase alphabets while referring to a service (for example: `s3`, `iam`, `ec2`, `sqs`, `sns`, `rds`, etc.)

**Comment by SaketaChalamchala** [comment]
Thanks @tanvipenumudy and @kerneltime for the review. I updated the regex and added more test cases.

**Comment by tanvipenumudy** [comment]
Thank you for the contribution @SaketaChalamchala. The changes look good to me!

**Comment by smengcl** [comment]
Hi @SaketaChalamchala , thanks for the patch.

In the `How was this patch tested?` section. For the older header, is there a better error message to be return to the user, rather than `User doesn't have the right to access this resource`? That sounds pretty misleading to me so users might try to diagnose in the wrong direction (they could think it is permission issue rather than the auth signature issue).

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/security/TestAWSV4AuthValidator.java [hunk 6]
```diff
@@ -64,15 +67,74 @@ public static Collection<Object[]> data() {
                 "577efef23edd43b7e1a59",
             "5d672d79c15b13162d9279b0855cfba" +
                 "6789a8edb4c82c400e06b5924a6f2b5d7",
-            "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY"
+            "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY",
+            true
+        },
+        // Invalid Algorithm
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/security/TestAWSV4AuthValidator.java:73]: New test case: invalid signature

```suggestion
        },
        // Invalid signature
        {
            "AWS4-HMAC-SHA256\n" +
                "20150830T123600Z\n" +
                "20150830/us-east-1/iam/aws4_request\n" +
                "f536975d06c0309214f805bb90ccff089219ecd68b2" +
                "577efef23edd43b7e1a59",
            "5d672d79c15b13162d9279b0855cfba" +
                "6789a8edb4c82c400e06b5924a6f2b5d8",
            "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY",
            false
        },
        // Invalid Algorithm
```

**Comment by SaketaChalamchala** [comment]
Closing this PR. Will submit another with some basic validations on the S3G side before we validate the signature.

**Pull request closed without merging**